### PR TITLE
[ci-app] Javascript – Add Known Limitations and Best Practices

### PR DESCRIPTION
**Pull request by juan-fernandez**
### What does this PR do?
* Add `Best practices` and `Known limitations` sections to the Javascript Tracer. 

### Motivation
* Set the correct expectations for users of CI Visibility with Javascript. 
* Define best practices for taking full advantage of the tool.  

### Preview

https://docs-staging.datadoghq.com/juan-fernandez/ci-visibility-add-known-limitations/continuous_integration/setup_tests/javascript

### Additional Notes

---

### Reviewer checklist
- [ ] Review the changed files.
- [ ] Review the URLs listed in the [Preview](#preview) section.
- [ ] Check images for PII
- [ ] Review any mentions of "Contact Datadog support" for internal support documentation.

### CODE REVIEW

File: content/en/continuous_integration/setup_tests/javascript.md [hunk 1]
```diff
@@ -15,8 +15,9 @@ further_reading:
 Supported test frameworks:
 * Jest >=24.8.0
   * Only `jsdom` (in package `jest-environment-jsdom`) and `node` (in package `jest-environment-node`) are supported as test environments. Custom environments like `@jest-runner/electron/environment` in `jest-electron-runner` are not supported.
-  * Only [`jest-circus`][1] is supported as a `testRunner`. 
+  * Only [`jest-circus`][1] is supported as a `testRunner`.
 * Mocha >=5.2.0
+  * Mocha >=9.0.0 has [**partial support**](#known-limitations).
```
Comment (kayayarai) [inline content/en/continuous_integration/setup_tests/javascript.md:20]:
```suggestion
  * Mocha >=9.0.0 has [partial support](#known-limitations).
```

File: content/en/continuous_integration/setup_tests/javascript.md [hunk 2]
```diff
@@ -120,6 +121,54 @@ Use different test scripts for CI and local development:
 },
 {{< /code-block >}}
 
+## Known limitations
+
+### ES Modules
+[Mocha >=9.0.0][8] has gone ESM-first to load test files. That means that if ES modules are used (for example, by defining test files with the .mjs extension), **the instrumentation will be limited**: tests will still be detected, but there won't be visibility within your test. If you want to read more about ES modules, you can do it [here][9].
```
Comment (kayayarai) [inline content/en/continuous_integration/setup_tests/javascript.md:127]:
```suggestion
[Mocha >=9.0.0][8] uses an ESM-first approach to load test files. That means that if ES modules are used (for example, by defining test files with the `.mjs` extension), _the instrumentation will be limited_. Tests are detected, but there isn't visibility into your test. For more information about ES modules, see [the NodeJS documentation][9].
```

File: content/en/continuous_integration/setup_tests/javascript.md [hunk 3]
```diff
@@ -120,6 +121,54 @@ Use different test scripts for CI and local development:
 },
 {{< /code-block >}}
 
+## Known limitations
+
+### ES Modules
```
Comment (kayayarai) [inline content/en/continuous_integration/setup_tests/javascript.md:126]:
```suggestion
### ES modules
```

File: content/en/continuous_integration/setup_tests/javascript.md [hunk 4]
```diff
@@ -120,6 +121,54 @@ Use different test scripts for CI and local development:
...
+[Mocha >=9.0.0][8] has gone ESM-first to load test files. That means that if ES modules are used (for example, by defining test files with the .mjs extension), **the instrumentation will be limited**: tests will still be detected, but there won't be visibility within your test. If you want to read more about ES modules, you can do it [here][9].
+
+### Browser tests
+The Javascript Tracer does not support browsers. This means that if you're running browser tests with `mocha` or `jest`, there won't be visibility within the test itself.
```
Comment (kayayarai) [inline content/en/continuous_integration/setup_tests/javascript.md:130]:
```suggestion
The JavaScript tracer does not support browsers, so if you run browser tests with `mocha` or `jest`, there isn't visibility within the test itself.
```

File: content/en/continuous_integration/setup_tests/javascript.md [hunk 5]
```diff
@@ -120,6 +121,54 @@ Use different test scripts for CI and local development:
...
+
+## Best practices
+
+The following practices allow to take full advantage of the testing framework and CI Visibility.
```
Comment (kayayarai) [inline content/en/continuous_integration/setup_tests/javascript.md:134]:
```suggestion
Follow these practices to take full advantage of the testing framework and CI Visibility.
```

File: content/en/continuous_integration/setup_tests/javascript.md [hunk 6]
```diff
@@ -120,6 +121,54 @@ Use different test scripts for CI and local development:
...
+
+
+### Parameterized tests
+
+Whenever possible, leverage the tools the testing frameworks provide for parameterized tests. For example, for `jest`:
```
Comment (kayayarai) [inline content/en/continuous_integration/setup_tests/javascript.md:139]:
```suggestion
Whenever possible, leverage the tools that testing frameworks provide for parameterized tests. For example, for `jest`:
```

File: content/en/continuous_integration/setup_tests/javascript.md [hunk 7]
```diff
@@ -120,6 +121,54 @@ Use different test scripts for CI and local development:
...
+Avoid this:
+{{< code-block lang="javascript" >}}
+[[1,2,3], [3,4,7]].forEach((a,b,expected) => {
+  test('sums correctly', () => {
+    expect(a+b).toEqual(expected)
+  })
+})
+{{< /code-block >}}
+
+And use instead the supported [`test.each`][10] instead:
```
Comment (kayayarai) [inline content/en/continuous_integration/setup_tests/javascript.md:150]:
```suggestion
And use instead the supported [`test.each`][10] instead:
```
```suggestion
And use [`test.each`][10] instead:
```

File: content/en/continuous_integration/setup_tests/javascript.md [hunk 8]
```diff
@@ -120,6 +121,54 @@ Use different test scripts for CI and local development:
...
+{{< code-block lang="javascript" >}}
+test.each([[1,2,3], [3,4,7]])('sums correctly %i and %i', (a,b,expected) => {
+  expect(a+b).toEqual(expected)
+})
+{{< /code-block >}}
+
+For `mocha`, [`mocha-each`][11] is supported:
```
Comment (kayayarai) [inline content/en/continuous_integration/setup_tests/javascript.md:157]:
```suggestion
For `mocha`, use [`mocha-each`][11]:
```

File: content/en/continuous_integration/setup_tests/javascript.md [hunk 9]
```diff
@@ -120,6 +121,54 @@ Use different test scripts for CI and local development:
...
+forEach([
+  [1,2,3],
+  [3,4,7]
+])
+.it('adds %i and %i then returns %i', (a,b,expected) => {
+  expect(a+b).to.equal(expected)
+});
+{{< /code-block >}}
+
+By using this approach, both the testing framework and CI Visibility will be able to tell your tests apart.
```
Comment (kayayarai) [inline content/en/continuous_integration/setup_tests/javascript.md:169]:
```suggestion
When you use this approach, both the testing framework and CI Visibility can tell your tests apart.
```

**Review by kayayarai — CHANGES_REQUESTED** [review]
Some grammar and style edits

**Comment by juan-fernandez** [comment]
thanks for the great feedback @kayayarai ! I applied all of it

**Review by kayayarai — APPROVED** [review]

**Pull request merged**